swift/ql/lib/codeql/swift/elements/expr/SelfRefExpr.qll

```diff
@@ -0,0 +1,14 @@
+private import codeql.swift.elements.expr.DeclRefExpr
+private import codeql.swift.elements.decl.MethodDecl
+private import codeql.swift.elements.decl.VarDecl
+
+/** A reference to `self`. */
+class SelfRefExpr extends DeclRefExpr {
+  MethodDecl methodDecl;
+
+  SelfRefExpr() { this.getDecl() = methodDecl.getSelfParam() }
```

Comment:
Hm, possibly. Can a MethodDecl contain other Decls as children? Since the SelfRefExpr could then appear in the body of the sub-Decl, and the identity you suggest wouldn't hold. But if not, then it's reasonable.

Reply:
Also, `getEnclosing{Function,Decl}` are final and cached in AstNode.qll, so probably best to introduce a separate predicate. As for the name, I wouldn't mind `getDeclaring(Method)Decl`, like I was planning on giving to Decls as part of the `IterableDeclContext` rename.

Reply:
Speaking of, maybe `DeclaringDecl` would be a good name for `IterableDeclContext`, moreso than `DeclWithMembers` :)

Reply:
> Hm, possibly. Can a MethodDecl contain other Decls as children? Since the SelfRefExpr could then appear in the body of the sub-Decl, and the identity you suggest wouldn't hold. But if not, then it's reasonable.

You mean declaring a `Decl` field inside the `MethodDecl` class? Sure, that's not a problem.

> Also, `getEnclosing{Function,Decl}` are final and cached in AstNode.qll, so probably best to introduce a separate predicate. As for the name, I wouldn't mind `getDeclaring(Method)Decl`, like I was planning on giving to Decls as part of the `IterableDeclContext` rename.

Ah, I see what you're saying: We do get the right behavior from the `getEnclosing{Function,Decl}` predicates that we get from extending `DeclRefExpr`, but you want to provide an override that has the correct return type, and you do that by creating a new predicate `getDeclaringMethod`. In that case, I think I think I'm happy with this change 👍